Basic curvefit implementation

slevang opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Closes #4300
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] New functions/methods are listed in `api.rst`

This is a simple implementation of a more general curve-fitting API as discussed in #4300, using the existing scipy `curve_fit` functionality wrapped with `apply_ufunc`. It works for arbitrary user-supplied 1D functions that ingest numpy arrays. Formatting and nomenclature of the outputs was largely copied from `.polyfit`, but could probably be improved.

pep8speaks commented:
Hello @slevang! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`xarray/core/dataarray.py`](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataarray.py):

> [Line 4292:14](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataarray.py#L4292): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ':'
> [Line 4305:86](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataarray.py#L4305): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 4308:71](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataarray.py#L4308): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 4310:86](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataarray.py#L4310): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 4311:88](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataarray.py#L4311): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 4328:85](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataarray.py#L4328): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`xarray/core/dataset.py`](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py):

> [Line 7013:14](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7013): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ':'
> [Line 7026:86](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7026): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 7029:71](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7029): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 7031:86](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7031): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 7032:88](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7032): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 7049:85](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7049): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 7079:9](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7079): [F401](https://duckduckgo.com/?q=pep8%20F401) '.computation.apply_ufunc' imported but unused
> [Line 7103:38](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7103): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 7117:17](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7117): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 7118:17](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7118): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 7119:17](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7119): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 7122:35](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7122): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 7122:44](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7122): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 7122:54](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7122): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 7123:44](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7123): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 7123:54](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7123): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 7124:44](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7124): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 7134:47](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7134): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 7134:63](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7134): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 7134:79](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7134): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 7137:20](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7137): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 7145:1](https://github.com/pydata/xarray/blob/e06027ae71134e2ae98c502d2104a8e233feacc6/xarray/core/dataset.py#L7145): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1

TomNicholas commented:
This is great, thanks for submitting this!

I just had a go with it, and it worked nicely. I have a couple of suggestions for improving it though:

1) Different fit coefficients as differently-named variables in the output, rather than indexed with a coordinate. This would then be consistent with `Dataset.polyfit`, which returns a set of different variables `[var]_polyfit_coefficients`. We could also get the names from the names of the keyword args when inspecting the function, and if it fails to get names just call them like `param1_fit_coefficients`, `param2_fit_coefficients` etc.

    What you have now works nicely though, so perhaps you could just reorganise the result before returning, like `Dataset.polyfit` does?

2) Initial guesses for each fit parameter. At the moment the user has to pass an ordered array of initial guesses through like

    ```python
    da.curvefit(x=da.x, dim='x', func=linear, kwargs={'p0': [m_guess, c_guess]})
    ```

    but it would be nicer to just pass them as a dictionary like

    ```python
    da.curvefit(x=da.x, dim='x', func=linear, initial_guess={'m': m_guess, 'c': c_guess})
    ```
    or even have the guesses read from the function definition maybe? i.e.
    ```python
    def linear(x, m=m_guess, c=c_guess):
        return m*x + c
    ```

3) (Stretch goal) Ability to fit >1D functions, e.g. fit a 2D gaussian to find a peak in a 2D image. But if we get the API right then this could be left to a later PR.

Also, the whole argument inspection thing probably deserves a few dedicated tests, in addition to testing the fitting functionality.

slevang commented:
> 1. Different fit coefficients as differently-named variables in the output, rather than indexed with a coordinate. 

I think the way I configured things now does replicate the polyfit results. For example:
```python
ds = xr.tutorial.open_dataset('air_temperature')
ds['air2'] = ds.air.copy()
ds.polyfit(dim='time', deg=2)

<xarray.Dataset>
Dimensions:                    (degree: 3, lat: 25, lon: 53)
Coordinates:
  * degree                     (degree) int64 2 1 0
  * lat                        (lat) float64 75.0 72.5 70.0 ... 20.0 17.5 15.0
  * lon                        (lon) float64 200.0 202.5 205.0 ... 327.5 330.0
Data variables:
    air_polyfit_coefficients   (degree, lat, lon) float64 -1.162e-32 ... 1.13...
    air2_polyfit_coefficients  (degree, lat, lon) float64 -1.162e-32 ... 1.14...
```
Compared to this:
```python
def square(x, a, b ,c):
    return a*np.power(x, 2) + b*x + c

ds.curvefit(x=ds.time, dim='time', func=square)

<xarray.Dataset>
Dimensions:                     (lat: 25, lon: 53, param: 3)
Coordinates:
  * lat                         (lat) float32 75.0 72.5 70.0 ... 20.0 17.5 15.0
  * lon                         (lon) float32 200.0 202.5 205.0 ... 327.5 330.0
  * param                       (param) <U1 'a' 'b' 'c'
Data variables:
    air_curvefit_coefficients   (param, lat, lon) float64 -1.162e-32 ... 1.13...
    air2_curvefit_coefficients  (param, lat, lon) float64 -1.162e-32 ... 1.14...
```
In both cases, each variable in the dataset returns a separate coefficients variable, and all fittable coefficients are stacked along a dimension, `degree` for the more specific polyfit case and a generic `param` for curvefit.

> 2. Initial guesses for each fit parameter.

Yeah this would be good. Should be easy to look for default values in the function itself using `inspect`.

> 3. (Stretch goal) Ability to fit >1D functions

Looks like this could be possible with a call to `ravel` like [this](https://scipython.com/blog/non-linear-least-squares-fitting-of-a-two-dimensional-data/). I'll do some experimenting.

TomNicholas commented:
> I think the way I configured things now does replicate the polyfit results.

You're right! My bad. The consistency with polyfit looks good.

> Looks like this could be possible with a call to ravel

Oh nice! That looks like it would allow for ND functions fit to ND data. It looks like there is a [dask version of ravel](https://docs.dask.org/en/latest/array-api.html#dask.array.ravel) which might be useful. (And judging by the comments on that blog post I think @StanczakDominik would appreciate this feature too!)

slevang commented:
Some more progress here.

1. I liked the idea of being able to supply `p0` and `bounds` as a dictionary mapping between parameter names and values, which is much more in the style of xarray. So I've included the logic to translate between this and the scipy args which are just ordered lists. The logic first prioritizes any of these values supplied directly in keyword args, then looks for defaults in the function definition, and otherwise falls back to the scipy defaults.
2. There's now a working implementation of multi-dimensional fits using `ravel` to flatten the coordinate arrays. One limitation is that `func` must be specified like `f((x, y), z)` as scipy takes a single `x` argument for the fitting coordinate. Maybe there is a clever way around this, but I think it's ok.
3. Some changes to the API. 

The best way to specify the fitting coordinates is a bit tricky to figure out. My original use case for this was needing to fit a relationship between two time/lat/lon dataarrays with the fit done over all time. But probably a more common use would be to just fit a curve over one or two dimensions that already exist in your data. So it would be great to handle these possibilities seamlessly. 

What I've settled on for now is a `coords` argument that takes a list of coordinates that should be the same length as the input coordinates of your fitting function. Additionally, there is a `reduce_dim` arg, so that, say if you want to fit a function in time but aggregate over all lat and lon, you can supply those dimensions here. So with a 3D x/y/time array, any of the following should work:
```python
# Fit a 1d function in time, returns parameters with dims (x, y)
da.curvefit(coords='time', ...)
# Fit a 2d function in space, returns parameters with dims (t)
da.curvefit(coords=['x', 'y'], ...)
# Fit a 1d function with another 3d dataarray and aggregate over time, returns parameters with dims (x, y)
da.curvefit(coords=da1, reduce_dim='time', ...)
```
The logic to make this work got a bit complicated, since we need to supply the right `input_core_dims` to `apply_ufunc`, and also to explicitly broadcast the coordinates to ensure the `ravel` operation works. Any thoughts on cleanup here appreciated.

Will eventually need to add tests and improve docs and examples. Tests especially I could use some help on.

slevang commented:
Added a couple usage examples in the docs, including one that replicates the [scipy](https://scipython.com/blog/non-linear-least-squares-fitting-of-a-two-dimensional-data/) example of fitting multiple peaks. Because of the wrapper function and variable args, this requires supplying `param_names` manually. Works nicely though.

Illviljan commented on xarray/core/dataset.py:
```diff
@@ -7007,5 +7008,222 @@ def argmax(self, dim=None, axis=None, **kwargs):
+        """
+        from scipy.optimize import curve_fit
+
+        from .alignment import broadcast
+        from .computation import apply_ufunc
+        from .dataarray import DataArray
```
```suggestion
```
xarray is already imported at the top can't you use xr.broadcast xr.apply_ufunc and xr.DataArray instead?

slevang replied:
Yep, you're right. Fixed in the latest commit.

slevang commented:
Added some checks that will raise errors if `inspect.signature` is not able to determine the function arguments, either for functions with variable-length args or things like numpy ufuncs that don't seem to work with `signature`. In these cases, `param_names` can be passed manually.

Also added minimal tests, but these should probably be expanded.

Illviljan commented on xarray/core/dataset.py:
```diff
@@ -7007,5 +7008,235 @@ def argmax(self, dim=None, axis=None, **kwargs):
+        self,
+        coords: Union["DataArray", Iterable["DataArray"]],
+        func: Callable[..., Any],
+        reduce_dim: Union[Hashable, Iterable[Hashable]] = None,
+        skipna: bool = True,
+        cov: bool = False,
+        p0: Dict[str, Any] = None,
+        bounds: Dict[str, Any] = None,
+        param_names: Sequence[str] = None,
+        kwargs: Dict[str, Any] = None,
```
```suggestion
        reduce_dim: Union[Hashable, Iterable[Hashable]] = [],
        skipna: bool = True,
        cov: bool = False,
        p0: Dict[str, Any] = {},
        bounds: Dict[str, Any] = {},
        param_names: Sequence[str] = None,
        kwargs: Dict[str, Any] = {},
```
Can't we have better default values instead to avoid all those None checks?

dcherian replied:
This is not a good idea: https://docs.python-guide.org/writing/gotchas/#what-you-wrote

slevang replied:
I defer to the experts. This is my first PR here, so was trying to copy the typing behavior from other functions.

slevang replied:
Also welcome any other feedback on proper type hinting. For example, I think we should actually allow:
```python
coords: Union[Union[str, "DataArray"], Iterable[Union[str, "DataArray"]]]
```
But this raises errors in mypy:
```
xarray/core/dataset.py:7108: error: Item "str" of "Union[str, DataArray]" has no attribute "equals"
xarray/core/dataset.py:7121: error: Item "str" of "Union[Any, str, DataArray]" has no attribute "broadcast_like"
```
Even though any strings passed to `coords` get converted to `self[coord]`.

Illviljan replied:
@dcherian That's interesting. It feels like I've used lists and dicts as defaults so many times and yet I don't remember this quirk. Maybe I just never run those functions more than once...

mathause replied:
mypy does not like the redefining `List[str] -> List[DataArray]` on line 7104. You'll need to rename the variable, e.g. `coords_` or we'd need to [add a mypy flag](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-allow-redefinition).

slevang replied:
Thought I had tried that approach (renaming `coords` in 7104), but that does indeed make mypy happy. Thanks.

slevang commented:
I've been playing around with this some more, and found the performance to be much better using a process-heavy dask scheduler. For example:
```python
import xarray as xr
import numpy as np
import time
import dask
​
def exponential(x, a, xc):
    return np.exp((x - xc) / a)
​
x = np.arange(-5, 5, 0.001)
t = np.arange(-5, 5, 0.01)
X, T = np.meshgrid(x, t)
Z1 = np.random.uniform(low=-5, high=5, size=X.shape)
Z2 = exponential(Z1, 3, X) + np.random.normal(scale=0.1, size=Z1.shape)
​
ds = xr.Dataset(
    data_vars=dict(var1=(["t", "x"], Z1), var2=(["t", "x"], Z2)),
    coords={"t": t, "x": x},
)
​
ds = ds.chunk({'x':10})
​
def test_fit():
    start = time.time()
    fit = ds.var2.curvefit(
        coords=ds.var1,
        func=exponential,
        reduce_dim="t",
    ).compute()
    print(f'Fitting time: {time.time() - start:.2f}s')

with dask.config.set(scheduler='threads'):
    test_fit()
with dask.config.set(scheduler='processes'):
    test_fit()
with dask.distributed.Client() as client:
    test_fit()
with dask.distributed.Client(n_workers=8, threads_per_worker=1) as client:
    test_fit()
```
On my 8-core machine, takes:
```
Fitting time: 8.32s
Fitting time: 2.71s
Fitting time: 4.40s
Fitting time: 3.43s
```
According to [this](https://github.com/scipy/scipy/pull/8740) the underlying scipy routines should be thread safe.

dcherian reviewed (commented):
Thanks @slevang This is an amazing first PR!

It's very thorough and nicely written. I have just minor comments. 

The only major comment is that I suggest refactoring some code out to a couple of helper functions which can then be tested independently. A few more test cases would be nice but I think you've covered most of the functionality.

slevang commented on xarray/core/dataset.py:
```diff
@@ -7007,5 +7008,235 @@ def argmax(self, dim=None, axis=None, **kwargs):
+        if n_params == 1:
+            result = result.squeeze(["param", "cov_i", "cov_j"])
+
+        if not cov:
```
No strong opinion here, but this replicates the `polyfit` behavior, which defaults to not returning the covariances. However, scipy returns them by default, so we aren't saving any computation by dropping them. I agree, would be fine to leave this to the user to deal with.

dcherian replied:
let's remove this then.

slevang commented on xarray/core/dataset.py:
```diff
@@ -7007,5 +7008,235 @@ def argmax(self, dim=None, axis=None, **kwargs):
+        result = result.assign_coords(
+            {"param": params, "cov_i": params, "cov_j": params}
+        )
+        result = result.transpose("param", "cov_i", "cov_j", ...)
```
Same as above, this replicates the `polyfit` behavior where `degree` is transposed to the first axis.

dcherian replied:
OK i think we should remove it. it doesn't really add anything

cc @aulemahal

aulemahal replied:
In `polyfit`, the behaviour is dictated by numpy. It wasn't on purpose, so I'd agree that transposing here is unnecessary.

slevang replied:
Sounds good, I'll remove it.

slevang commented on xarray/core/dataset.py:
```diff
@@ -7007,5 +7008,235 @@ def argmax(self, dim=None, axis=None, **kwargs):
+        result = xr.Dataset()
+        cov_fields = ["cov_i", "cov_j"]
+        for name, da in self.data_vars.items():
+            if isinstance(name, str):
```
I don't have a good understanding of how the `ReprObject` works, and again the previous code was copied right from `polyfit`. If I set `_THIS_ARRAY` within the method (which I think is your suggestion), I get equality but not identity with `name`. However, this seems to work:
```python
if name is xr.core.dataarray._THIS_ARRAY:
    name = ""
else:
    name = f"{str(name)}_"
```

dcherian replied:
OK lets do this. thanks!

slevang commented on xarray/core/dataset.py:
```diff
@@ -7007,5 +7008,235 @@ def argmax(self, dim=None, axis=None, **kwargs):
+            return popt, pcov
+
+        result = xr.Dataset()
+        cov_fields = ["cov_i", "cov_j"]
```
Only to drop the covariance matrix and associated dimensions later on. Can be removed if we eliminate the `cov` kwarg.

slevang commented on xarray/core/dataset.py:
```diff
@@ -7007,5 +7008,235 @@ def argmax(self, dim=None, axis=None, **kwargs):
+                    "`param_names` must be provided because `func` takes variable length arguments."
+                )
+
+        n_params = len(params)
```
Sounds like a good idea. Where should these live?

slevang replied:
Ok. I guess they could either be module-level private functions at the top of `dataset.py` (but I can't imagine them being used elsewhere), or we could put them within the `Dataset` class with a `@staticmethod` decorator (this also seems a bit funky). 

Refactoring things out into `fitting.py` sounds like a good long term solution either way.

dcherian replied:
private functions sounds fine to me.

I didn't realize this was in `dataset.py` Another option is to move it to `computation.py`.

slevang replied:
`polyfit` is in `dataset.py`, and I do think it makes sense to call it as a dataset method, since a common use case would be just fitting a function along a dimension. But factoring out most of the code to functions in `computation.py` or a new `fitting.py` would help keep things organized.

dcherian replied:
I think we add private methods for now and refactor later.

slevang commented:
Thanks for the review @dcherian! The latest commit has been refactored with a couple helper functions and associated tests, and any steps that served no purpose other than consistency with `polyfit` have been removed.

If you can think of any more specific test cases that should be included, happy to add them.

slevang commented:
I think so. I pushed a merge commit to get this up to date with the current release.

dcherian commented:
Thanks @slevang. Sorry for the delay!
